### PR TITLE
Turn on strict typing in mypy

### PR DESCRIPTION
**Pull request by Avasam**
- Pinned mypy to ensure the CI won't start failing suddenly if next release finds new issues
- Bumped `python_version` to 3.8 to stop warnings about mypy not supporting Python 3.7 (relates to #2207 )
- Updated and improved comments
- Start checking `setup.py`
- Solved about a dozen of type issues

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -965,24 +969,17 @@ def my_new_compiler(**kw):
...
 
-
 # No way to cleanly wedge our compiler sub-class in.
-from distutils import ccompiler
-from distutils._msvccompiler import MSVCCompiler
-
 orig_new_compiler = ccompiler.new_compiler
-ccompiler.new_compiler = my_new_compiler
-
-base_compiler = MSVCCompiler
```
Comment (Avasam) [inline setup.py:976]: Doesn't look like `base_compiler` was used for anything other than being an alias.

**Comment by Avasam** [comment]
@mhammond When you have the time, can you turn on autofixing in https://results.pre-commit.ci/ ? Direct link: https://github.com/apps/pre-commit-ci/installations/new

File: mypy.ini [hunk 2]
```diff
@@ -1,30 +1,46 @@
...
-; Pythonwin/Scintilla is vendored
-; Pythonwin/pywin/idle is vendored IDLE extensions predating Python 2.3. They now live in idlelib in https://github.com/python/cpython/tree/main/Lib/idlelib
-; Ignoring non-public apis for now
-; Duplicate module named "rasutil" and "setup", short-term fix is to ignore
-exclude = .*((build|adodbapi|Pythonwin/Scintilla|Pythonwin/pywin/idle|[Tt]est|[Dd]emos?)/.*|rasutil.py|setup.py)
+disable_error_code =
+    ; Module has no attribute; (Dynamic modules will be hard to type without first-party stubs, ie: pythoncom.*)
+    attr-defined,
+    ; Class cannot subclass "..." (has type "Any"); (IDEM)
+    ; TODO: Use typeshed's types-pywin32 stubs after a few more fixes there
```
Comment (Avasam) [inline mypy.ini:23]: https://github.com/python/typeshed/pull/11616

File: mypy.ini [hunk 3]
```diff
@@ -1,30 +1,46 @@
...
+    misc,
```
Comment (Avasam) [inline mypy.ini:24]: https://github.com/python/typeshed/pull/11616

File: win32/Lib/win32timezone.py [hunk 4]
```diff
@@ -907,7 +908,8 @@ def resolveMUITimeZone(spec):
 
 
 # from jaraco.util.dictlib 5.3.1
-class RangeMap(dict):
+# TODO: Update to implementation in jaraco.collections
```
Comment (Avasam) [inline win32/Lib/win32timezone.py:911]: https://github.com/jaraco/jaraco.util/commit/e00b92f17130d3bedc3a88e324eeabcf440e84ff

File: setup.py [hunk 5]
```diff
@@ -2298,7 +2296,7 @@ def maybe_fixup_exes():
         long_description = f.read()
     long_description_content_type = "text/markdown"
 
-dist = setup(
+dist = setup(  # type: ignore[func-returns-value] # FIXME: Pending fix in typeshed
```
Comment (Avasam) [inline setup.py:2299]: https://github.com/python/typeshed/pull/11617

File: mypy.ini [hunk 6]
```diff
@@ -36,3 +52,7 @@ ignore_missing_imports = True
 ; pywin32_system32 is an empty module created in setup.py to store dlls
 [mypy-verstamp,win32com.*,Test,pywin32_system32]
 ignore_missing_imports = True
+
+; See https://github.com/python/typeshed/pull/10948
+[mypy-distutils]
+ignore_missing_imports = True
```
Comment (Avasam) [inline mypy.ini:58]: https://github.com/python/typeshed/pull/10948

**Review by mhammond — APPROVED** [review]

**Pull request merged**